Fix #7166: Use WebKit Session Restore (InteractionState)

Brandon-T opened the pull request:
## Summary of Changes
- Add SessionRestore from WebKit
- Store state on background and terminate :)
- Update database version.

<!-- Enter a ticket number for this PR, create a new one if it is not there yet. -->
This pull request fixes #7166

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Test Plan:
- Test Session Restore
- Test restoring reader-mode
- Test restoring error pages
- Test recently closed


## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue and pull request is assigned to a milestone (should happen at merge time).

Brandon-T committed:
Fix restoring internal tabs
Fix tab saving scroll position.

Fix recently closed showing internal URLs.
Fix so that reader-mode now shows a favicon! :)

Signed-off-by: Brandon T <JustBrandonT@gmail.com>

iccub reviewed (commented):
Code looks good to me. Not approving yet, running the branch now looking for functional bugs

Brandon-T committed:
Removed the scroll observer. Scroll position is saved when the app terminates `cleanly` (does NOT happen on the simulator if you press Stop in Xcode. Only works when swiping up in the simulator and killing the app that way.

Signed-off-by: Brandon T <JustBrandonT@gmail.com>

soner-yuksel reviewed (commented):
I have some minor comments and questions. But this is great making a great baseline for multi window and tab groups using brand new API, great job

Brandon-T committed:
Store state on background and terminate :)
Add SessionRestore from WebKit

Update database version.
Migrate Tab screenshots to CoreData

Signed-off-by: Brandon T <JustBrandonT@gmail.com>
